core_java/twitter/src/main/java/ca/jrvs/apps/twitter/service/TwitterService.java

```diff
@@ -0,0 +1,166 @@
+package ca.jrvs.apps.twitter.service;
+
+import ca.jrvs.apps.twitter.dao.TwitterDao;
+import ca.jrvs.apps.twitter.model.GeoLoc;
+import ca.jrvs.apps.twitter.model.Tweet;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.text.Normalizer;
+import java.text.Normalizer.Form;
+import java.util.ArrayList;
+import java.util.List;
+import org.slf4j.LoggerFactory;
+
+/**
+ * TwitterService contains validation and some additional processing of Tweets being made through
+ * TwitterCLI.
+ */
+public class TwitterService implements Service {
+
+  private TwitterDao dao;
+
+  //Default Constructor
+  public TwitterService() {
+    dao = new TwitterDao();
+  }
+
+  //Alternative constructor for providing a custom TwitterDao
+  public TwitterService(TwitterDao dao) {
+    this.dao = dao;
+  }
+
+  /**
+   * Validate and post a user input Tweet.
+   *
+   * @param tweet tweet to be created
+   * @return created tweet
+   * @throws IllegalArgumentException if text exceed max number of allowed characters or lat/long
+   *                                  out of range
+   */
+  @Override
+  public Tweet postTweet(Tweet tweet) throws IllegalArgumentException {
+    String tweetText = tweet.getText();
+    GeoLoc coords = tweet.getLocation();
+    float latitude;
+    float longitude;
+    boolean validCoords = false;
+    boolean hasCoords = false;
+    boolean validText = false;
+
+    validText = validateText(tweetText);
+    if (coords != null) {
+      longitude = coords.getCoordinates()[0];
+      latitude = coords.getCoordinates()[1];
+      hasCoords = true;
+      validCoords = validateCoordinates(latitude, longitude);
+    }
+
+    if (validText) {
+      if (hasCoords && validCoords) {
```

Comment:
Don't think the if-else is necessary. Also if !hasCoords but !validCoords, would return null Tweet.  Think better handle would be to throw IllegalArugmentException

Reply:
Ok, just noticed validCoords is always going to return true or will throw exception. Think its good to clean it up a little - is confusing on first glance

Reply:
Changed so that validateCoords only returns true/false, the IllegalArg now throws in postTweet when checking the booleans (if has coords, but coods are invalid, throw IllegalArg)